### PR TITLE
Replicated Sharing Tensor: Multiplication, Matrix multiplication, and Convolution

### PR DESCRIPTION
**Pull request by abogaziah**
## Description
The implementation of the rest of the operations of replicated shared tensor that's necessary to implement falcon protocol.
1- Multiplication: to multiplay a secret x = [x1,x2,x3] by a secret y = [y1,y2,y3] : 

- Parties locally compute z1 = x1y1+x2y1+x1y2, z2 = x2y2+x3y2+x2y3 and z3 = x3y3 +x1y3 +x3y1.
At the end of this, z1,z2, and z3 form a 3-out-of-3 secret sharing of z

- Each party adds 3of3 correlated random number (alpha i) to zi where i is the index of the worker.

- Parties perform resharing by sending a copy of zi to party i+1

2- Matrix multiplication: multiplication algorithm works with matrix multiplication, just by replacing multiplication operator with matrix multiplication

3-  Convolution: convolution is done by converting to one matrix multiplication as explained [here](https://cs231n.github.io/convolutional-networks/)  using [torch unfold](https://discuss.pytorch.org/t/custom-convolution-dot-product/14992/27) 

4- Applied black box principle, i.e. the class should take in a tensor and a set of workers and output a tensor, all other details should be abstracted inside the class, I did that by removing shares_map from the parameters and change other detail around it  

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests
______________________________________________________
- [x] Multiplication #3900 
- [x] Matrix multiplication #3901 
- [x] convolution #3901

### CODE REVIEW

File: syft/frameworks/torch/mpc/przs.py
Comment (gmuraru) [inline syft/frameworks/torch/mpc/przs.py:8]: Q: should this be 5?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:219]: I really like this function :D

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:39]: Q: ```)``` should be after ```(plain_text - sum(shares))```

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:263]: Q: Is this used?

File: syft/frameworks/torch/mpc/przs.py
Comment (abogaziah) [reply syft/frameworks/torch/mpc/przs.py:8]: it should be less than or equal the ring size of RST

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:219]: yeah me too :D it saved a lot of code repetition

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:39]: yes it should

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:263]: No, I thought I will need it for convolution, but I didn't have to do any in-place operations, should I remove it? 
I think I should remove it, but I'm like 99% sure I'll need it in the future and I'll regret removing it 😂

Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py]: removed it

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:33]: Q: Shouldn't we return ```self``` here?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:14]: Q: Shouldn't we do this only if we have ```plain_text != None```?
(maybe we can have two ways in which we construct RSS, one when we send the plaintext to the constructor and one using the ```share``` method)

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:18]: Q: Shouldn't we better raise an exception here? (maybe something like ```BadValue```)

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:48]: Q: In case plain_text is a ```float``` this will simply do a "cast" to an ```int``` type, right?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:195]: Q: Should we have this in the ```RSS``` or in the ```Falcon``` class?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:69]: Q: Can we make this ```__retrieve_shares```?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:76]: Q: Also this one, ``` __retrieve_pointers?```

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:256]: Q: This is used only internally. Should it be ```public```?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:33]: No, not anymore, this function was used to kinda init the object but this doesn't happen anymore

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:14]: this is checked in the __generate_shares_map function. if plain_text's None it returns None

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:18]: sometimes you need to make an empty tensor like an empty torch tensor, it should be acceptable I think

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:48]: it will cast it from any torch.dtype to torch.long

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:195]: hmm, it could be implemented both ways, but the paper described it as an operation of the sharing scheme, so let's stick to the paper for now, it can be changed easily later if it is inconvenient

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:69]: I meant to make them public since they are very useful for debugging to see the shares before adding them, it's basically the reconstruction algorithm but you'll have to add the shares manually

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:256]: any reshaping operation like view, unfold, reshape (or any other non-arithmetic operation) will use this function, so it will be used externally

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py
Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:12]: Hmm...seems ok - when refactoring we should stick with the same name method (we have ```share``` for ```AST``` and we do the sharing in the constructor for the ```RST```)

**Review by gmuraru — APPROVED** [review]

**Comment by gmuraru** [comment]
LGTM! After you merge it, could you give me a ping to refactor [this](https://github.com/OpenMined/PySyft/pull/4474)

**Pull request merged**
